eureka2-client/src/main/java/com/netflix/eureka2/client/interest/BatchingRegistryImpl.java

```diff
@@ -0,0 +1,170 @@
+import com.netflix.eureka2.interests.StreamStateNotification;
+import com.netflix.eureka2.interests.StreamStateNotification.BufferingState;
+import com.netflix.eureka2.utils.rx.PauseableSubject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Notification;
+import rx.Observable;
+import rx.Observable.OnSubscribe;
+import rx.Subscriber;
+import rx.functions.Action0;
+import rx.functions.Func1;
+import rx.subjects.PublishSubject;
+
+/**
+ * @author Tomasz Bak
+ */
+public class BatchingRegistryImpl<T> implements BatchingRegistry<T> {
+
+    private static final Logger logger = LoggerFactory.getLogger(BatchingRegistryImpl.class);
+
+    private final Map<Interest<T>, BufferingState> interestsBufferingState =
+            new ConcurrentHashMap<Interest<T>, BufferingState>();
+
+    private final PauseableSubject<Boolean> updatesSubject = PauseableSubject.create();
+
+    /**
+     * As channels may reconnect, and this object is scoped to an {@link com.netflix.eureka2.interests.IndexRegistry}
+     * we need to handle multiple subscriptions from subsequent channels, with state cleanup.
+     * If the change notification stream from channel terminates, clean the state, and start over from the
+     * new channel.
+     */
+    private final PublishSubject<Observable<ChangeNotification<T>>> notificationSources = PublishSubject.create();
+
+    public BatchingRegistryImpl() {
+        notificationSources.switchMap(
+                new Func1<Observable<ChangeNotification<T>>, Observable<Boolean>>() {
+                    @Override
+                    public Observable<Boolean> call(Observable<ChangeNotification<T>> notifications) {
+                        return notifications.map(new Func1<ChangeNotification<T>, Boolean>() {
+                            @Override
+                            public Boolean call(ChangeNotification<T> notification) {
+                                if (notification instanceof StreamStateNotification) {
+                                    StreamStateNotification<T> stateNotification = (StreamStateNotification<T>) notification;
+                                    interestsBufferingState.put(stateNotification.getInterest(), stateNotification.getBufferingState());
+                                    return true;
+                                }
+                                return null;
+                            }
+                        }).filter(new Func1<Boolean, Boolean>() {
+                            @Override
+                            public Boolean call(Boolean value) {
+                                return value != null;
+                            }
+                        }).materialize().flatMap(new Func1<Notification<Boolean>, Observable<Boolean>>() {
+                            @Override
+                            public Observable<Boolean> call(Notification<Boolean> materialized) {
+                                switch (materialized.getKind()) {
```

Comment:
OnError/OnComplete have the same logic here, that is outside of the switch statement. I will add it anyway so it is more explicit in the code.